File: day_01/day01_b.lua
local file = 'input.txt'

-- list
local a_list = {}
local b_list = {}

for line in io.lines(file) do
    local a, b = string.match(line, '(%d+)%s+(%d+)')
    table.insert(a_list, a)
    table.insert(b_list, b)
end

local count_table = {}
for _, entry in pairs(b_list) do
    local index = entry
    if count_table[index] == nil then
        count_table[index] = 1
    else
        count_table[index] = count_table[index] + 1
    end
end

print("Counted 📋: ")
for k, v in pairs(count_table) do
    print(k .. " --- ".. v)
end

local result = 0
for _, entry in pairs(a_list) do
    if count_table[entry] == nil then
        
    else
        result = result + entry * count_table[entry]
    end
end
print(result)

File: day_01/day01_b.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, _ := os.Open("input.txt")
	fscanner := bufio.NewScanner(file)
	var a_list []int64
	var b_list []int64
	for fscanner.Scan() {
		pairs := strings.Split(fscanner.Text(), "   ")

		a, _ := strconv.ParseInt(pairs[0], 10, 64)
		b, _ := strconv.ParseInt(pairs[1], 10, 64)
		a_list = append(a_list, a)
		b_list = append(b_list, b)
	}

	lookup := make(map[int64]int64)
	for _, value := range b_list {

		if _, exists := lookup[value]; exists {
			lookup[value] += 1
		} else {
			lookup[value] = 1
		}
	}

	var result int64 = 0
	for _, value := range a_list {
		mult_val := lookup[value]
		result = result + value*mult_val
	}

	fmt.Println(result)
}
